designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+#### Approach #2
+
+This is how [kustomize](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/) discovers plugins.
+The PROJECT file in Kubebuilder has `plugins` field and each plugin will have the resources: Group, Kind and Version, so every plugin defined in the PROJECT file configuration can get its own GVK directory for the executable to be placed in and Kubebuilder will search for a plugin with the name of the plugin in the GVK directory of the plugin. Once KB successfully locates the plugin, it will run the plugin using the CLI.
+
+Following `kustomize`'s way of placement, every plugin gets its own directory such as:
+
+```project
+    $XDG_CONFIG_HOME/kubebuilder/plugins
+    /${group}/${version}/LOWERCASE(${kind}) 
+```
+
+The default value of XDG_CONFIG_HOME is $HOME/.config.
+
+* Pros
+  * `kustomize` which is popular and robust tool, follows this approach in which `apiVersion` and `kind` fields are used to locate the plugin.
+
+  * This appoach does not enforce naming of plugins but enforces the directories the plugins have to be in, which is a more common and typical approach.
+
+  * The one-plugin-per-directory requirement eases creation of a plugin bundle for sharing.
+
+* Cons
+  * This makes it mandatory for every plugin that is specified in the `layout` field to have additional configuration information such as Group, Version and Kind in the `plugins` field in the PROJECT file for it to be discoverable. Currently, the `plugins` field is more like an optional configuration storage field.
+
+#### Approach #3
+
+Another approach I would like to mention is adding plugin executables with a prefix `kubebuilder-` followed by the plugin name to the PATH variable. This will enable KB to traverse through the PATH looking for the plugin executables starting with the prefix `kubebuilder-` and matching by the plugin name that was provided in the CLI. Furthermore, a check should be added to verify that the match is an executable or not and return an error if it's not an executable. This approach provides a lot of flexibility in terms of plugin discovery as all the user needs to do is to add the plugin executable to the PATH and Kubebuilder will discover it.
+
+* Pros
+  * `kubectl` and `git` follow the same approach for discovering plugins, so there’s prior art.
+
+  * There’s a lot of flexibility in just dropping plugin binaries to PATH variable and enabling the discovery without having to enforce any other constraints on the placements of the plugins.
+
+* Cons
+  * Enumerating the list of all available plugins might be a bit tough compared to having a single folder with the list of available plugins and having to enumerate those.
+
+  * These plugin binaries cannot be run in a standalone manner outside of Kubebuilder, so may not be very ideal to add them to the PATH var.
```

Comment:
Is that true? Nothing should prevent someone from making their plugin act like its own program if they so choose. The only restriction should be that the plugin respect the contract that we specify i.e. STDIN/STDOUT some format of information that goes between that. Other than that, a plugin is just another executable that could do whatever it wants.